toolkit/components/uniffi-bindgen-gecko-js/src/templates/js/macros.jsm

```diff
@@ -6,8 +6,19 @@
 {% call _call_scaffolding_function(cons, Some(object_type), "") %}
 {%- endmacro %}
 
-{%- macro call_method(method, object_type) %}
-{% call _call_scaffolding_function(method, method.return_type(), object_type.ffi_converter()) %}
+{%- macro call_method(method, object_type, object) %}
+    if (this.destroyed) {
+        throw new UniFFIError("Attempting to call method {{ method.nm() }} on Object that is already destroyed")
+    }
+    try {
+        this.callCounter += 1;
+        {% call _call_scaffolding_function(method, method.return_type(), object_type.ffi_converter()) %}
+    } finally {
+        this.callCounter -=1;
+        if (this.destroyed && this.callCounter === 0) {
+            return {{ ci.scaffolding_name() }}.{{ object.ffi_object_free().nm() }}(this.ptr);
+        }
+    }
```

Comment:
I think I'd prefer if the wrapper code here was in a function in `Object.jsm` -- either a method on the object or a top-level function.  I think it's easier to follow the logic if these macros focus on the function call and the `Object.jsm` code handles the destruction logic.